### PR TITLE
Tidy up SpecialProperties utils

### PR DESCRIPTION
**Pull request by bbrks**
- Renamed `stripSpecialSpecialProperties` -> `stripAllSpecialProperties`
- Reformatted ifs for clarity
- Pre-allocate new body capacity based on maximum possible size (no stripped properties)
```
21:20 $ go test -run=- -bench='BenchmarkStripSpecialProperties' -benchmem -count=3 ./db > old.txt
21:22 $ go test -run=- -bench='BenchmarkStripSpecialProperties' -benchmem -count=3 ./db > new.txt
21:23 $ benchcmp -best old.txt new.txt
benchmark                                         old ns/op     new ns/op     delta
BenchmarkStripSpecialProperties/no_special-12     1657          903           -45.50%
BenchmarkStripSpecialProperties/special-12        1689          944           -44.11%

benchmark                                         old allocs     new allocs     delta
BenchmarkStripSpecialProperties/no_special-12     4              2              -50.00%
BenchmarkStripSpecialProperties/special-12        4              2              -50.00%

benchmark                                         old bytes     new bytes     delta
BenchmarkStripSpecialProperties/no_special-12     2143          1202          -43.91%
BenchmarkStripSpecialProperties/special-12        2143          1202          -43.91%
```

### CODE REVIEW

File: db/revision.go [hunk 1]
```diff
@@ -425,31 +425,49 @@ func compareRevIDs(id1, id2 string) int {
...
+	return stripSpecialPropertiesExcept(b, nil)
+}
+
+// stripSpecialPropertiesExcept returns a copy of the given body with all underscore-prefixed keys removed, except those given.
+func stripSpecialPropertiesExcept(b Body, exceptions []string) Body {
+	// Assume no properties removed for the initial capacity to reduce allocs on large docs.
+	stripped := make(Body, len(b))
+	for k, v := range b {
+		if k == "" || k[0] != '_' ||
+			base.StringSliceContains(exceptions, k) {
```
Comment (adamcfraser) [inline db/revision.go:444]: Is there non-zero overhead associated with calling StringSliceContains for each key when exceptions is empty?  (i.e. can we add a case to the benchmark for stripAllSpecialProperties?)

File: db/revision.go [hunk 2]
```diff
@@ -425,31 +425,49 @@ func compareRevIDs(id1, id2 string) int {
...
-		if key != "" && key[0] == '_' && key != BodyId && key != BodyRev && key != BodyDeleted && key != BodyAttachments && key != BodyRevisions {
+// containsUserSpecialProperties returns true if the given body contains a non-SG special property (underscore prefixed)
+func containsUserSpecialProperties(b Body) bool {
+	for k := range b {
+		if k != "" && k[0] == '_' &&
+			k != BodyId &&
+			k != BodyRev &&
+			k != BodyDeleted &&
+			k != BodyAttachments &&
+			k != BodyRevisions {
```
Comment (adamcfraser) [inline db/revision.go]: If it does turn out to have low overhead, should we use !base.StringSliceContains here?

**Review by adamcfraser — CHANGES_REQUESTED** [review]
Couple of questions/clarifications

File: db/revision.go (hunk 1 above)
Comment (bbrks) [reply db/revision.go:444]: Negligable difference. Updated benchmark to include `stripAllSpecialProperties`

```
BenchmarkStripSpecialProperties/no_specialAllSpecial-12         	 1289540	       925 ns/op	    1202 B/op	       2 allocs/op
BenchmarkStripSpecialProperties/no_specialAllSpecial-12         	 1293424	       929 ns/op	    1202 B/op	       2 allocs/op
BenchmarkStripSpecialProperties/no_specialAllSpecial-12         	 1282030	       926 ns/op	    1202 B/op	       2 allocs/op

BenchmarkStripSpecialProperties/no_specialOldSpecialSpecial-12  	 1291028	       927 ns/op	    1202 B/op	       2 allocs/op
BenchmarkStripSpecialProperties/no_specialOldSpecialSpecial-12  	 1293970	       930 ns/op	    1202 B/op	       2 allocs/op
BenchmarkStripSpecialProperties/no_specialOldSpecialSpecial-12  	 1267436	       932 ns/op	    1202 B/op	       2 allocs/op

BenchmarkStripSpecialProperties/specialAllSpecial-12            	 1301296	       924 ns/op	    1202 B/op	       2 allocs/op
BenchmarkStripSpecialProperties/specialAllSpecial-12            	 1303908	       921 ns/op	    1202 B/op	       2 allocs/op
BenchmarkStripSpecialProperties/specialAllSpecial-12            	 1299585	       922 ns/op	    1202 B/op	       2 allocs/op

BenchmarkStripSpecialProperties/specialOldSpecialSpecial-12     	 1305183	       916 ns/op	    1202 B/op	       2 allocs/op
BenchmarkStripSpecialProperties/specialOldSpecialSpecial-12     	 1309207	       915 ns/op	    1202 B/op	       2 allocs/op
BenchmarkStripSpecialProperties/specialOldSpecialSpecial-12     	 1307422	       920 ns/op	    1202 B/op	       2 allocs/op
```

File: db/revision.go (hunk 2 above)
Comment (bbrks) [reply db/revision.go]: Agreed

**Review by adamcfraser — APPROVED** [review]

**Pull request merged**